File: src/function_driven.rs
use std::{thread::sleep, time::Duration};

const CLEAR: &str = "\x1B[2J\x1B[1;1H";

fn progress<Iter>(iter: Iter, f: fn(&Iter::Item))
where Iter: Iterator {
    let mut i = 1;
    for n in iter {
        println!("{}{}", CLEAR, "*".repeat(i));
        i += 1;
        f(&n);
    }
}

fn expensive_calculation(_n: &i32) {
    sleep(Duration::from_secs(1));
}

fn main() {
    let v = vec![1, 2, 3];
    progress(v.into_iter(), expensive_calculation);
}

File: src/functionDriven.ts
const CLEAR = "\x1B[2J\x1B[1;1H";

function progress<T>(iter: Iterable<T>, f: (item: T) => void): void {
    let i = 1;
    for (const n of iter) {
        console.log(`${CLEAR}${"*".repeat(i)}`);
        i++;
        f(n);
    }
}

function expensiveCalculation(_n: number): void {
    // Simulate work with a blocking operation
    const start = Date.now();
    while (Date.now() - start < 1000) {
        // Busy wait for 1 second
    }
}

function main(): void {
    const v = [1, 2, 3];
    progress(v, expensiveCalculation);

    const b = "a b".split('').map(c => c.charCodeAt(0));
    progress(b, expensiveCalculation);
}


main();

export { 
    progress, 
    expensiveCalculation,
    main
};
